wdl-engine/src/http.rs

```diff
@@ -300,59 +321,111 @@ impl Downloader for HttpDownloader {
             // This loop exists so that all requests to download the same URL will block
             // waiting for a notification that the download has completed.
             // When the notification is received, the lookup into the downloads is retried
-            let mut retried = false;
+            let mut retried_wait = false;
             loop {
-                // Scope to ensure the mutex guard is not visible to the await point
-                let notify = {
+                let status_check_result = {
                     let mut downloads = self.downloads.lock().expect("failed to lock downloads");
                     match downloads.get(&url) {
                         Some(Status::Downloading(notify)) => {
                             assert!(
-                                !retried,
+                                !retried_wait,
                                 "file should not be downloading again after a notification"
                             );
-
-                            notify.clone()
+                            Either::Left(notify.clone())
                         }
                         Some(Status::Downloaded(r)) => {
                             return r.clone();
                         }
                         None => {
-                            assert!(
-                                !retried,
-                                "file should not be downloaded again after a notification"
-                            );
-
-                            // Insert an entry and break out of the loop to download
+                            // not downloading, not downloaded. mark for retry/download
                             downloads.insert(
                                 url.clone().into_owned(),
-                                Status::Downloading(Arc::default()),
+                                Status::Downloading(Arc::new(Notify::new())),
                             );
-                            break;
```

Comment:
Instead of this `Either`, you should just use an `Option`, as you use it exclusively as an option (logically speaking) below.